### PR TITLE
update ascent package with recent ver dep logic, and dray support

### PR DESCRIPTION
**Pull request by cyrush**
depends on updates from PRs:

#17501
#17500 
#17499
#17498 
#17497 
#17496 
#17497 
#17496 
#17495 
#17494 

(split up as requested by adam)

### CODE REVIEW

**Comment by cyrush** [comment]
Split up -- several packages now failing unit tests b/c they depend now depend on packages from other PRs, which aren't there.  What is the best way to detangle this? Do you want a merge order?

**Comment by ax3l** [comment]
Suggested merge order would be fab, yes :) but we can just take a look which ones pass and you rebase the dependent ones after merge.

Do you want to add yourself to `maintainers = ['cyrush']` of those new packages? :)

File: var/spack/repos/builtin/packages/ascent/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ascent/package.py:46]: Why remove the stable release? We usually like to have at least one stable release that gets installed by default since develop is a moving target.

File: var/spack/repos/builtin/packages/ascent/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ascent/package.py]: So this isn't needed for `+babelflow~mpi`? I think your `spec['babelflow'].prefix` below will crash in this case.

File: var/spack/repos/builtin/packages/ascent/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/ascent/package.py:154]: Is there any way we can simplify this variant forwarding mess? I know the new concretizer is still in development, but this is getting out of hand...

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/ascent/package.py
Comment (cyrush) [reply var/spack/repos/builtin/packages/ascent/package.py]: thanks for noticing this, I moved babelflow logic to be mpi dependent.

File: var/spack/repos/builtin/packages/ascent/package.py
Comment (cyrush) [reply var/spack/repos/builtin/packages/ascent/package.py:46]: stable release will no longer build after these changes to spack. Also we are not supporting earlier releases, given the pace at which bug fixes and features are being put in.

File: var/spack/repos/builtin/packages/ascent/package.py
Comment (cyrush) [reply var/spack/repos/builtin/packages/ascent/package.py:154]: We aren't a fan of having to leverage this much logic, that said:

- without this plumbing (and that in dependent packages), an out-of-the box build (for E4S for example) simply wont work. While there are other mechanisms to apply constraints for build (spack envs?), those mechanisms aren't employed in important cases -- we are aiming to keep out of the box working. I suspect we will need to increase the pace of our spack develop updates.

- supporting both cuda and non-cuda imposes restrictions on static vs shared libs, and certain features. We need to plumb these or else builds will fail. This is super important to us.

- we want to migrate our development process to use a newer spack develop to build our tpls. Our current process uses a fork from 11/19. We need this plumbing to support our automated builds and the install process we use for folks on Summit, etc.

I hope we can simplify in the future, but for now we need to persevere in the face of current limitations.

File: var/spack/repos/builtin/packages/ascent/package.py
Comment (cyrush) [reply var/spack/repos/builtin/packages/ascent/package.py:46]: also on this topic: Making a stable release work require another pass at combos of variants and versions, exploding the depends on logic more.  Also, we have to use a custom version of VTK-m. This dep has lots of API changes, we aren't able to use tagged releases or develop, we have to dial a specific version in. VTK-m folks are ok with us setting a custom version for use in ascent in their spack package, but they don't want these to start piling up.

**Review by adamjstewart — APPROVED** [review]

**Comment by cyrush** [comment]
Awesome -- thanks for all of  your help on these!

**Pull request merged**
